Task4_0603 train.csv 파일로 생존자 예측 모델 생성을 위한 분석용 데이터 셋을 작성하세요.

In [124]:
import pandas as pd
from google.colab import files
uploaded = files.upload()

# 업로드된 파일을 데이터프레임으로 읽기
file_path = list(uploaded.keys())[0]
df = pd.read_csv(file_path)

Saving train.csv to train (19).csv


In [125]:
df.groupby(['Sex','Survived'])['Survived'].count()

Sex     Survived
female  0            81
        1           233
male    0           468
        1           109
Name: Survived, dtype: int64

In [126]:
df[['Pclass', 'Survived']].groupby('Pclass').mean()

Survived
Pclass          
1       0.629630
2       0.472826
3       0.242363

In [127]:
df[['Sex','Pclass','Survived']].groupby(['Sex', 'Pclass']).mean()

Survived
Sex    Pclass          
female 1       0.968085
       2       0.921053
       3       0.500000
male   1       0.368852
       2       0.157407
       3       0.135447

In [128]:
df.groupby(['Sex', 'Pclass'])['Survived'].mean()

Sex     Pclass
female  1         0.968085
        2         0.921053
        3         0.500000
male    1         0.368852
        2         0.157407
        3         0.135447
Name: Survived, dtype: float64

In [129]:
df.groupby(['Embarked', 'Pclass'])['Survived'].mean()

Embarked  Pclass
C         1         0.694118
          2         0.529412
          3         0.378788
Q         1         0.500000
          2         0.666667
          3         0.375000
S         1         0.582677
          2         0.463415
          3         0.189802
Name: Survived, dtype: float64

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [ ]:
df_numeric = df.select_dtypes(include='number')
df_numeric.corr()['Survived']

PassengerId   -0.005007
Survived       1.000000
Pclass        -0.338481
Age           -0.077221
SibSp         -0.035322
Parch          0.081629
Fare           0.257307
Name: Survived, dtype: float64

In [113]:
# 필요없는 열 제거
# df.drop(columns = ['PassengerId','Name','Ticket','Cabin','Pclass','Embarked'], inplace=True)
df.drop(columns = ['PassengerId','Name','Ticket','Cabin','Fare'], inplace=True)
df.head()

Survived  Pclass     Sex   Age  SibSp  Parch Embarked
0         0       3    male  22.0      1      0        S
1         1       1  female  38.0      1      0        C
2         1       3  female  26.0      0      0        S
3         1       1  female  35.0      1      0        S
4         0       3    male  35.0      0      0        S

In [27]:
df.drop(columns=['Pclass'],inplace=True)

In [4]:
# 결측치 확인
print(df.isnull().sum())

Survived      0
Pclass        0
Sex           0
Age         177
SibSp         0
Parch         0
Fare          0
Embarked      2
dtype: int64


In [14]:
df.Age.value_counts()

Age
24.00    30
22.00    27
18.00    26
19.00    25
28.00    25
         ..
36.50     1
55.50     1
0.92      1
23.50     1
74.00     1
Name: count, Length: 88, dtype: int64

In [114]:
df['Embarked'].fillna(df['Embarked'].mode()[0], inplace = True)

In [103]:
df.Age.fillna(df.Age.mode()[0],inplace=True)

In [5]:
df.isnull().sum()

Survived    0
Pclass      0
Sex         0
Age         0
SibSp       0
Parch       0
Embarked    0
dtype: int64

In [115]:
# 결측치 제거
print(df.isnull().sum())
df.Age.fillna(df.Age.mean(), inplace=True)
print(df.isnull().sum())

Survived      0
Pclass        0
Sex           0
Age         177
SibSp         0
Parch         0
Embarked      0
dtype: int64
Survived    0
Pclass      0
Sex         0
Age         0
SibSp       0
Parch       0
Embarked    0
dtype: int64


In [116]:
# 파생변수 생성
df['family_size'] = df['SibSp']+df['Parch']+1
df.head()

Survived  Pclass     Sex   Age  SibSp  Parch Embarked  family_size
0         0       3    male  22.0      1      0        S            2
1         1       1  female  38.0      1      0        C            2
2         1       3  female  26.0      0      0        S            1
3         1       1  female  35.0      1      0        S            2
4         0       3    male  35.0      0      0        S            1

In [117]:
# 필요없는 열 제거
df.drop(columns=['SibSp','Parch'],inplace=True)
df.head()

Survived  Pclass     Sex   Age Embarked  family_size
0         0       3    male  22.0        S            2
1         1       1  female  38.0        C            2
2         1       3  female  26.0        S            1
3         1       1  female  35.0        S            2
4         0       3    male  35.0        S            1

In [68]:
df

Survived  Pclass     Sex   Age Embarked  family_size
0           0       3    male  22.0        S            2
1           1       1  female  38.0        C            2
2           1       3  female  26.0        S            1
3           1       1  female  35.0        S            2
4           0       3    male  35.0        S            1
..        ...     ...     ...   ...      ...          ...
886         0       2    male  27.0        S            1
887         1       1  female  19.0        S            1
888         0       3  female  24.0        S            4
889         1       1    male  26.0        C            1
890         0       3    male  32.0        Q            1

[891 rows x 6 columns]

In [119]:
df.head(10)

Survived  Pclass     Sex        Age Embarked  family_size  \
0         0       3    male  22.000000        S            2   
1         1       1  female  38.000000        C            2   
2         1       3  female  26.000000        S            1   
3         1       1  female  35.000000        S            2   
4         0       3    male  35.000000        S            1   
5         0       3    male  29.699118        Q            1   
6         0       1    male  54.000000        S            1   
7         0       3    male   2.000000        S            5   
8         1       3  female  27.000000        S            3   
9         1       2  female  14.000000        C            2   

   pclass_by_sex_age  
0          27.372153  
1          34.141405  
2          24.068493  
3          34.141405  
4          27.372153  
5          27.372153  
6          39.287717  
7          27.372153  
8          24.068493  
9          28.748661

In [106]:
# 클래스별 나이 평균
df['pclass_by_age'] = df.groupby('Pclass').Age.transform(lambda x : x.mean())

In [107]:
# age가 성인일때 1 미성년자일 때 0
df['adult'] = df['Age'].apply(lambda x: 1 if x >18 else 0)

In [28]:
# 성별에 따른 나이 평균
df['sex_mean'] = df.groupby('Sex').Age.transform(lambda x : x.mean())

In [118]:
# 클래스별 성별에 따른 나이 평균
df['pclass_by_sex_age'] = df.groupby(['Pclass','Sex']).Age.transform(lambda x : x.mean())

In [120]:
# age_group을 6개 범주로 확대
bins = [0,5,12,18,35,60,100]
labels = ['Infant','Child','Teenager','Yound Adult','Adult','Senior']
df['age_group'] = pd.cut(df['Age'], bins=bins, labels = labels)
df.drop(columns=['Age'], inplace = True)
df.head()

Survived  Pclass     Sex Embarked  family_size  pclass_by_sex_age  \
0         0       3    male        S            2          27.372153   
1         1       1  female        C            2          34.141405   
2         1       3  female        S            1          24.068493   
3         1       1  female        S            2          34.141405   
4         0       3    male        S            1          27.372153   

     age_group  
0  Yound Adult  
1        Adult  
2  Yound Adult  
3  Yound Adult  
4  Yound Adult

In [121]:
category_column = ['age_group','Sex','Embarked','Pclass','family_size','pclass_by_sex_age']

for column in category_column:
    df= pd.get_dummies(df,columns=[column])


In [122]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
# 타겟 변수와 피처 변수 설정
X = df.drop('Survived', axis=1)
y = df['Survived']

# 학습 데이터와 테스트 데이터 분할
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 데이터 크기 확인
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

# 랜덤 포레스트 모델 학습
model = RandomForestClassifier(n_estimators=100, random_state=42)
model.fit(X_train, y_train)

# 예측
y_pred = model.predict(X_test)

# 성능 평가
print("Accuracy:", accuracy_score(y_test, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("Classification Report:\n", classification_report(y_test, y_pred))


(712, 29) (179, 29) (712,) (179,)
Accuracy: 0.8324022346368715
Confusion Matrix:
 [[97  8]
 [22 52]]
Classification Report:
               precision    recall  f1-score   support

           0       0.82      0.92      0.87       105
           1       0.87      0.70      0.78        74

    accuracy                           0.83       179
   macro avg       0.84      0.81      0.82       179
weighted avg       0.84      0.83      0.83       179



- 생존과 상관없거나 null이 많은 열은 제거
- fare과 pclass를 같이 넣었을 때, 각 각 넣어봤을 때
 가장 높게 나온건 pclass survived와 상관계수를 확인해봤을 때 pclass가 더 높아서 그런 것 같다
- SibSp, Parch를 각 각 넣은 것보다 파생 변수를 만드는게 0.1정도 높다
- age의 결측값을 평균값이 아니라 최빈값으로 넣었을 때 예측도가 0.01정도 높아진다
- age와 관련된 파생 변수를 넣을 때 조금이라도 높아진다

결측값이 발생한 이유가 그 범주의 대표값이었을 경우라면, 최빈값으로 대체하였을 때 데이터의 분포를 더 정확하게 보존할 수 있기 때문입니다.

결측값이 그 변수의 중요한 특성을 반영하고 있다는 것을 의미할 수 있다.

평균값으로
Classification Report:
               precision    recall  f1-score   support

           0       0.81      0.91      0.86       105
           1       0.85      0.70      0.77        74

    accuracy                           0.83       179

   macro avg       0.83      0.81      0.82       179

weighted avg       0.83      0.83      0.82       179

평균값 대신 최빈값으로

Classification Report:
               precision    recall  f1-score   support

           0       0.82      0.92      0.87       105
           1       0.87      0.70      0.78        74

    accuracy                           0.83       179

   macro avg       0.84      0.81      0.82       179

weighted avg       0.84      0.83      0.83       179

성인인지 아닌지 파생변수 포함

Classification Report:
               precision    recall  f1-score   support

           0       0.82      0.91      0.86       105
           1       0.85      0.72      0.78        74

    accuracy                           0.83       179

   macro avg       0.84      0.82      0.82       179

weighted avg       0.83      0.83      0.83       179



성인, 클래스별 나이 평균 파생변수 포함(age의 결측값을 최빈값, 평균값을 넣어도 같다)

Classification Report:
               precision    recall  f1-score   support

           0       0.82      0.92      0.87       105
           1       0.87      0.70      0.78        74

    accuracy                           0.83       179

   macro avg       0.84      0.81      0.82       179

weighted avg       0.84      0.83      0.83       179
